torch/csrc/jit/codegen/cuda/python_frontend/fusion_manager.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+#include <c10/macros/Export.h>
+
+#include <torch/csrc/jit/codegen/cuda/kernel_cache.h>
+#include <torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h>
+
+#include <memory>
+
+//! nvFuser Fusion IR namespace abbreviation
+namespace Nvf = torch::jit::fuser::cuda;
+
+namespace nvfuser {
+
+struct RecordFunctor;
+
+//! \struct FusionCacheEntry
+//! \brief Is the container for a Node in the cache contained in the
+//! FusionManager that is organized as a prefix tree.
+
+struct TORCH_CUDA_CU_API FusionCacheEntry {
+  FusionCacheEntry(std::shared_ptr<RecordFunctor>& rec);
+  FusionCacheEntry();
+
+  //! An entry's primary data is the record it holds
+  std::shared_ptr<RecordFunctor> record;
+  //! A hash map of the children for the current node.
+  //! The hash map hashs a pointer to a RecordFunctor because
+  //! the hash function is virtual.  Also, a shared_ptr is used
+  //! to own the pointer because the original owner of a pointer,
+  //! a FusionDefintion, can't be relied upon to hold
+  //! hold the record over the lifetime of the FusionManager.
+  std::unordered_map<
+      std::shared_ptr<RecordFunctor>,
+      std::unique_ptr<FusionCacheEntry>>
+      record_hash_map;
+
+  //! This boolean indicates a leaf node with a cached nvFuser Fusion
+  bool is_terminal;
+  //! A pointer to the nvFuser object that holds an unscheduled Fusion
+  std::unique_ptr<Nvf::FusionExecutorCache> fusion_executor_cache;
+};
+
+//! \class FusionManager
+//! \brief A singleton class used in the nvFuser python interface
+//! to manage the caching of fusions.
+//!
+//! Example:
+//!
+//!   fm = FusionManager.get()
```

Comment:
Looking at this part of the example, one thing I'm not sure is how fusion is cached across different fusion groups.

So `FusionManager.get()` gives us access to the singleton that holds the cache tree. At the exit of `FusionDefinition(fm) as fd`, we traverse fm to point to the active FusionCacheEntry.

Does this mean that FusionDefinition have to be redefined for each following iteration during cache hit? Maybe I'm missing something here, since I don't see FusionCAcheEntry exposed at pybind to be cached somewhere.

Reply:
^^^ @kevinstephano This is currently the part that I'm not getting from looking at the PR. Would like to sort it out first.

Reply:
I think we discussed this live.